grpc/src/main/java/com/linecorp/armeria/server/grpc/UnframedGrpcErrorHandlerBuilder.java

```diff
@@ -0,0 +1,221 @@
+        requireNonNull(responseTypes, "responseTypes");
+        if (this.responseTypes == null) {
+            this.responseTypes = EnumSet.noneOf(UnframedGrpcErrorResponseType.class);
+        }
+        this.responseTypes.addAll(ImmutableSet.copyOf(responseTypes));
+        return this;
+    }
+
+    /**
+     * Registers custom messages to be marshalled by the error handler.
+     *
+     * <p>This method registers specific message instances for custom error responses.
+     * If a custom JSON marshaller has already been set, calling this method will
+     * result in an {@link IllegalArgumentException}.
+     *
+     * @param messages The message instances to register
+     */
+    public UnframedGrpcErrorHandlerBuilder registerMarshalledMessages(Message... messages) {
+        requireNonNull(messages, "messages");
+        if (jsonMarshaller != null) {
+            throw new IllegalArgumentException(
+                    "Cannot register custom messages because a custom JSON marshaller has already been set. " +
+                    "Use the custom marshaller to register custom messages.");
+        }
+        this.messages.addAll(ImmutableList.copyOf(messages));
+        return this;
+    }
+
+    /**
+     * Registers custom message types to be marshalled by the error handler.
+     *
+     * <p>This method registers specific message types for custom error responses.
+     * If a custom JSON marshaller has already been set, calling this method will
+     * result in an {@link IllegalArgumentException}.
+     *
+     * @param messageTypes The message types to register
+     */
+    @SafeVarargs
+    public final UnframedGrpcErrorHandlerBuilder registerMarshalledMessages(
+            Class<? extends Message>... messageTypes) {
+        requireNonNull(messageTypes, "messageTypes");
+        if (jsonMarshaller != null) {
+            throw new IllegalArgumentException(
+                    "Cannot register custom messageTypes because a custom JSON marshaller has already been " +
+                    "set. Use the custom marshaller to register custom message types.");
+        }
+        this.messageTypes.addAll(ImmutableList.copyOf(messageTypes));
+        return this;
+    }
+
+    /**
+     * Registers custom messages or message types to be marshalled by the error handler.
+     *
+     * <p>This method allows registering either message instances or message types for
+     * custom error responses. Each element in {@code messagesOrMessageTypes} must either
+     * be an instance of the {@link Message} class　or a class that extends the {@link Message}
+     * class. If a custom JSON marshaller has already been set,
+     * calling this method will result in an {@link IllegalArgumentException}.
+     *
+     * @param messagesOrMessageTypes The messages or message types to register
+     */
+    public UnframedGrpcErrorHandlerBuilder registerMarshalledMessages(Iterable<?> messagesOrMessageTypes) {
```

Comment:
What do you think about splitting `registerMarshalledMessages` into two?

- `registerMarshalledMessages()`
- `registerMarshalledMessageTypes()`

This way, we don't need dynamic type inspection because we will have:

- `registerMarshalledMessages(Iterable<? extends Message>)` and
- `registerMarshalledMessageTypes(Iterable<? extends Class<? extends Message>>)`

Reply:
I thought we might consider that a user wants to register a list containing both `Message` instances and `Message` extended classes. Do you think that is rare case?

Reply:
It was rather because of Java generics' limitation. We cannot define the following two methods because they are actually same after type erasure:

- `register(Iterable<? extends Message>)`
- `register(Iterable<? extends Class<? extends Message>>)`

If we change their names so they have different names, we don't need to worry about this language limitation.